File: mobile/android/app/src/main/java/com/dots/herness/mobile/McpClient.kt
package com.dots.herness.mobile

import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import org.json.JSONArray
import org.json.JSONObject
import java.io.File
import java.net.HttpURLConnection
import java.net.URL
import java.util.UUID

/**
 * MCP over streamable HTTP. Only HTTPS transports are accepted from a phone in
 * practice: a stdio server needs a child process and a binary to run, which an
 * installed app has no way to provide for arbitrary servers.
 */
/** The token is held in the keystore, never in this record or the servers file. */
data class McpServer(val id: String = UUID.randomUUID().toString(), val name: String, val url: String, val token: String = "")

class McpClient(private val server: McpServer) {
    private var sessionId: String? = null
    private var nextId = 0

    suspend fun listTools(): List<Triple<String, String, JSONObject>> {
        call("initialize", JSONObject()
            .put("protocolVersion", "2025-06-18")
            .put("capabilities", JSONObject())
            .put("clientInfo", JSONObject().put("name", "HerNess mobile").put("version", "1.0")))
        runCatching { notify("notifications/initialized") }
        val tools = call("tools/list", JSONObject()).optJSONArray("tools") ?: JSONArray()
        return (0 until tools.length()).mapNotNull { index ->
            val tool = tools.optJSONObject(index) ?: return@mapNotNull null
            val name = tool.optString("name").ifBlank { return@mapNotNull null }
            Triple(
                name,
                tool.optString("description").ifBlank { "MCP tool $name from ${server.name}." },
                tool.optJSONObject("inputSchema") ?: JSONObject().put("type", "object"),
            )
        }
    }

    suspend fun callTool(name: String, arguments: JSONObject): String {
        val result = call("tools/call", JSONObject().put("name", name).put("arguments", arguments))
        val content = result.optJSONArray("content") ?: JSONArray()
        val text = (0 until content.length()).mapNotNull { index ->
            val block = content.optJSONObject(index) ?: return@mapNotNull null
            when (block.optString("type")) {
                "text" -> block.optString("text")
                "resource" -> block.optJSONObject("resource")?.optString("text") ?: "[resource]"
                else -> "[${block.optString("type", "content")}]"
            }
        }.joinToString("\n")
        if (result.optBoolean("isError")) error(text.ifBlank { "The MCP tool reported an error." })
        return text.ifBlank { "The MCP tool returned no content." }
    }

    private suspend fun call(method: String, params: JSONObject): JSONObject {
        nextId += 1
        val body = JSONObject().put("jsonrpc", "2.0").put("id", nextId).put("method", method).put("params", params)
        val response = decode(send(body))
        response.optJSONObject("error")?.let { error(it.optString("message", "The MCP server reported an error.")) }
        return response.optJSONObject("result") ?: JSONObject()
    }

    private suspend fun notify(method: String) {
        send(JSONObject().put("jsonrpc", "2.0").put("method", method).put("params", JSONObject()))
    }

    private suspend fun send(body: JSONObject): String = withContext(Dispatchers.IO) {
        val parsed = runCatching { java.net.URI(server.url.trim()) }.getOrNull()
        check(parsed?.scheme?.lowercase() == "https" && !parsed.host.isNullOrBlank() && parsed.userInfo == null) { "HTTPS MCP endpoint required" }
        val connection = (URL(server.url).openConnection() as HttpURLConnection).apply {
            requestMethod = "POST"
            connectTimeout = 15_000
            readTimeout = 120_000
            doOutput = true
            setRequestProperty("Content-Type", "application/json")
            setRequestProperty("Accept", "application/json, text/event-stream")
            if (server.token.isNotBlank()) setRequestProperty("Authorization", "Bearer ${server.token}")
            sessionId?.let { setRequestProperty("Mcp-Session-Id", it) }
            outputStream.use { it.write(body.toString().toByteArray()) }
        }
        val status = connection.responseCode
        connection.getHeaderField("Mcp-Session-Id")?.let { sessionId = it }
        val text = (if (status in 200..299) connection.inputStream else connection.errorStream)?.use { String(it.readBytes()) } ?: ""
        if (status !in 200..299) error("${server.name}: $status $text")
        text
    }

    /** A streamable-HTTP server answers with plain JSON or an SSE stream whose last `data:` frame carries the response. */
    private fun decode(text: String): JSONObject {
        runCatching { return JSONObject(text) }
        text.lineSequence().filter { it.startsWith("data:") }.toList().reversed().forEach { line ->
            runCatching { return JSONObject(line.removePrefix("data:").trim()) }
        }
        error("${server.name} returned a response that is not JSON-RPC.")
    }
}

/** The servers the phone knows about, and the agent tools they expose. */
class McpRegistry(private val storage: File, private val secure: SecureStore) {
    var servers by mutableStateOf<List<McpServer>>(emptyList()); private set
    var status by mutableStateOf<Map<String, String>>(emptyMap()); private set

    init { load() }

    fun add(name: String, url: String, token: String) {
        val server = McpServer(name = name.ifBlank { url }, url = url)
        if (token.isNotBlank()) secure.write("mcp.${server.id}", token)
        servers = servers + server
        save()
    }

    fun remove(server: McpServer) {
        secure.delete("mcp.${server.id}")
        servers = servers.filterNot { it.id == server.id }
        status = status - server.id
        save()
    }

    /** Connects to every server and hands the discovered tools to the agent, namespaced by server. */
    suspend fun connectAll(agent: MobileAgent) {
        servers.forEach { server ->
            val client = McpClient(server.copy(token = secure.read("mcp.${server.id}").orEmpty()))
            try {
                val tools = client.listTools()
                agent.register(tools.map { (name, description, schema) ->
                    AgentToolSpec(toolName(server, name), description, schema) { input -> client.callTool(name, input) }
                })
                status = status + (server.id to "${tools.size} tool(s)")
            } catch (t: Throwable) {
                status = status + (server.id to (t.message ?: "Connection failed."))
            }
        }
    }

    private fun toolName(server: McpServer, tool: String) =
        "mcp__${server.name.lowercase().replace(Regex("[^a-z0-9_]"), "_")}__$tool"

    private fun save() {
        storage.parentFile?.mkdirs()
        storage.writeText(JSONArray(servers.map {
            JSONObject().put("id", it.id).put("name", it.name).put("url", it.url)
        }).toString())
    }

    private fun load() {
        if (!storage.exists()) return
        runCatching {
            val array = JSONArray(storage.readText())
            servers = (0 until array.length()).map { index ->
                val item = array.getJSONObject(index)
                McpServer(item.optString("id"), item.optString("name"), item.optString("url"))
            }
        }
    }
}

File: mobile/ios/Sources/MCPClient.swift
import Foundation

/// MCP over streamable HTTP. Only HTTPS transports are accepted from a phone:
/// stdio servers need a child process, which iOS does not allow, so a server has
/// to be something the phone can call over the network.
struct MCPServer: Codable, Identifiable, Equatable, Sendable {
    var id = UUID().uuidString
    var name: String
    var url: String
    /// Sent as `Authorization: Bearer …` when set. Stored in the keychain, not here.
    var usesToken = false

    var tokenKey: String { "mcp.\(id)" }
}

struct MCPTool: Sendable, Equatable {
    let name: String
    let description: String
    let schema: Data
}

actor MCPClient {
    private let server: MCPServer
    private let token: String
    private var sessionId: String?
    private var nextId = 0

    init(server: MCPServer, token: String) {
        self.server = server
        self.token = token
    }

    func listTools() async throws -> [MCPTool] {
        _ = try await call(method: "initialize", params: [
            "protocolVersion": "2025-06-18",
            "capabilities": [:],
            "clientInfo": ["name": "HerNess mobile", "version": "1.0"],
        ])
        try await notify(method: "notifications/initialized")
        let result = try await call(method: "tools/list", params: [:])
        let tools = (result["tools"] as? [[String: Any]]) ?? []
        return tools.compactMap { tool in
            guard let name = tool["name"] as? String else { return nil }
            let schema = tool["inputSchema"] as? [String: Any] ?? ["type": "object"]
            return MCPTool(
                name: name,
                description: tool["description"] as? String ?? "MCP tool \(name) from \(server.name).",
                schema: (try? JSONSerialization.data(withJSONObject: schema)) ?? Data("{\"type\":\"object\"}".utf8))
        }
    }

    func callTool(_ name: String, arguments: [String: Any]) async throws -> String {
        let result = try await call(method: "tools/call", params: ["name": name, "arguments": arguments])
        let content = result["content"] as? [[String: Any]] ?? []
        let text = content.compactMap { block -> String? in
            switch block["type"] as? String {
            case "text": return block["text"] as? String
            case "resource": return ((block["resource"] as? [String: Any])?["text"] as? String) ?? "[resource]"
            default: return "[\(block["type"] as? String ?? "content")]"
            }
        }.joined(separator: "\n")
        if result["isError"] as? Bool == true { throw MCPError.server(text.isEmpty ? "The MCP tool reported an error." : text) }
        return text.isEmpty ? "The MCP tool returned no content." : text
    }

    // MARK: - JSON-RPC

    private func call(method: String, params: [String: Any]) async throws -> [String: Any] {
        nextId += 1
        let body: [String: Any] = ["jsonrpc": "2.0", "id": nextId, "method": method, "params": params]
        let data = try await send(body: body)
        guard let object = decode(data) else { throw MCPError.server("\(server.name) returned a response that is not JSON-RPC.") }
        if let error = object["error"] as? [String: Any] {
            throw MCPError.server(error["message"] as? String ?? "The MCP server reported an error.")
        }
        return object["result"] as? [String: Any] ?? [:]
    }

    private func notify(method: String) async throws {
        _ = try? await send(body: ["jsonrpc": "2.0", "method": method, "params": [:]])
    }

    private func send(body: [String: Any]) async throws -> Data {
        guard let url = URL(string: server.url), url.scheme?.lowercased() == "https", url.user == nil, url.password == nil, url.host != nil else { throw MCPError.server("\(server.name) requires an HTTPS URL.") }
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.setValue("application/json, text/event-stream", forHTTPHeaderField: "Accept")
        if !token.isEmpty { request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization") }
        if let sessionId { request.setValue(sessionId, forHTTPHeaderField: "Mcp-Session-Id") }
        request.httpBody = try JSONSerialization.data(withJSONObject: body)
        request.timeoutInterval = 120

        let (data, response) = try await URLSession.shared.data(for: request)
        guard let http = response as? HTTPURLResponse else { throw MCPError.server("\(server.name) did not respond.") }
        if let value = http.value(forHTTPHeaderField: "Mcp-Session-Id") { sessionId = value }
        guard (200..<300).contains(http.statusCode) else {
            throw MCPError.server("\(server.name): \(http.statusCode) \(String(data: data, encoding: .utf8) ?? "")")
        }
        return data
    }

    /// A streamable-HTTP server may answer with a plain JSON body or with an SSE
    /// stream whose last `data:` frame carries the response.
    private func decode(_ data: Data) -> [String: Any]? {
        if let object = try? JSONSerialization.jsonObject(with: data) as? [String: Any] { return object }
        guard let text = String(data: data, encoding: .utf8) else { return nil }
        for line in text.split(separator: "\n").reversed() where line.hasPrefix("data:") {
            let payload = line.dropFirst(5).trimmingCharacters(in: .whitespaces)
            if let object = try? JSONSerialization.jsonObject(with: Data(payload.utf8)) as? [String: Any] { return object }
        }
        return nil
    }
}

enum MCPError: LocalizedError {
    case server(String)
    var errorDescription: String? { switch self { case .server(let message): return message } }
}

/// The servers the phone knows about, and the agent tools they expose.
@MainActor
final class MCPRegistry: ObservableObject {
    @Published private(set) var servers: [MCPServer] = []
    @Published private(set) var status: [String: String] = [:]

    private let secure = SecureStore()
    private static let storageKey = "herness.mcp.servers"

    init() {
        if let data = UserDefaults.standard.data(forKey: Self.storageKey),
           let value = try? JSONDecoder().decode([MCPServer].self, from: data) { servers = value }
    }

    func add(name: String, url: String, token: String) {
        var server = MCPServer(name: name.isEmpty ? url : name, url: url, usesToken: !token.isEmpty)
        if !token.isEmpty { secure.write(token, key: server.tokenKey) } else { server.usesToken = false }
        servers.append(server)
        save()
    }

    func remove(_ server: MCPServer) {
        secure.delete(server.tokenKey)
        servers.removeAll { $0.id == server.id }
        status[server.id] = nil
        save()
    }

    /// Connects to every server and hands the discovered tools to the agent,
    /// namespaced so two servers can expose the same tool name.
    func connectAll(into agent: MobileAgent) async {
        for server in servers {
            let client = MCPClient(server: server, token: server.usesToken ? (secure.read(server.tokenKey) ?? "") : "")
            do {
                let tools = try await client.listTools()
                let specs = tools.map { tool in
                    AgentToolSpec(
                        name: Self.toolName(server: server, tool: tool.name),
                        description: tool.description,
                        schema: (try? JSONSerialization.jsonObject(with: tool.schema)) as? [String: Any] ?? ["type": "object"]
                    ) { input in
                        let arguments = (try? JSONSerialization.jsonObject(with: input.data)) as? [String: Any] ?? [:]
                        return try await client.callTool(tool.name, arguments: arguments)
                    }
                }
                agent.register(specs)
                status[server.id] = "\(tools.count) tool(s)"
            } catch {
                status[server.id] = error.localizedDescription
            }
        }
    }

    static func toolName(server: MCPServer, tool: String) -> String {
        let slug = server.name.lowercased().replacingOccurrences(of: "[^a-z0-9_]", with: "_", options: .regularExpression)
        return "mcp__\(slug)__\(tool)"
    }

    private func save() {
        guard let data = try? JSONEncoder().encode(servers) else { return }
        UserDefaults.standard.set(data, forKey: Self.storageKey)
    }
}
